compiler/optimizer/IDT.cpp

```diff
@@ -0,0 +1,199 @@
+#include "IDT.hpp"
+
+#define INITIAL_NUM_children (10)
+
+IDTNode::IDTNode() {}
+
+IDTNode::IDTNode(IDT* idt, int32_t callsite_bci, TR::ResolvedMethodSymbol* rms)
+  : _idx(idt->next_idx())
+  , _callsite_bci(callsite_bci)
+  , _children(nullptr)
+  , _rms(rms)
+{}
+
+
+IDT::IDT(TR_InlinerBase* inliner, TR::Region *mem, TR::ResolvedMethodSymbol* rms)
+  : _mem(mem)
+  , _inliner(inliner)
+  , _max_idx(-1)
+  , _root(new (*_mem) IDTNode(this, -1, rms))
+{}
+
+void
+IDT::printTrace()
+{
+  if (comp()->trace(OMR::selectInliner)) {
+    TR_VerboseLog::writeLineLocked(TR_Vlog_SIP, "#IDT: %d candidate methods to inline into %s",
+      size() - 1,
+      getName(getRoot())
+    );
+    if (size() == 0) {
+      return;
+    }
+    getRoot()->printNodeThenChildren(this, -2);
+  }
+}
+
+void
+IDTNode::printNodeThenChildren(IDT* idt, int callerIndex)
+  {
+  if (this != idt->getRoot()) {
+    const char *nodeName = idt->getName(this);
+    TR_VerboseLog::writeLineLocked(TR_Vlog_SIP, "#IDT: %d: %d inlinable @%d -> bcsz=%d %s", 
+      _idx,
+      callerIndex,
+      _callsite_bci,
+      getBcSz(),
+      nodeName
+    );
+  }
+  
+  if (_children == NULL) {
+    return;
+  }
+
+  // Print children
+  IDTNode* child = IDT::getOnlyChild(this);
+
+  if (child != nullptr)
+    {
+    child->printNodeThenChildren(idt, _idx);
+    } 
+  else
+    {
+    for (auto curr = _children->begin(); curr != _children->end(); ++curr)
+      {
+      curr->printNodeThenChildren(idt, _idx);
+      }
+    }
+  }
+
+uint32_t
+IDTNode::getBcSz() {
+  return _rms->getResolvedMethod()->maxBytecodeIndex();
+}
+
+IDTNode*
+IDT::getOnlyChild(IDTNode* owner)
+  {
+  if (((size_t)owner->_children) & 1)
+    {
+    return (IDTNode *)(void*)(((size_t)(owner->_children)>>1)<<1);
+    }
+  return nullptr;
+  }
+
+void
+IDT::setOnlyChild(IDTNode* owner, IDTNode* child)
+  {
+  owner->_children = (IDTNode::Children*)(void*)((size_t)child | 1);
```

Comment:
the 1 should be pulled out as a define to give it semantic meaning:
`#define SINGLE_CHILD_BIT 1`
and then
`(IDTNode::Children*)(void*)((size_t)child | SINGLE_CHILD_BIT)`
here you want to cast child to uintptr_t not size_t since size_t is not going to be equal to a pointer width datatype. You should then be able to directly cast to IDTNod::Children